sway-core/src/query_engine/mod.rs

```diff
@@ -1,4 +1,54 @@
+use std::path::PathBuf;
+use std::sync::RwLock;
+use std::time::SystemTime;
+use std::{collections::HashMap, sync::Arc};
+
+use crate::{CompileResult, Programs};
+
+pub type ModulePath = Arc<PathBuf>;
+
+#[derive(Clone, Debug)]
+pub struct ModuleCacheEntry {
+    pub path: ModulePath,
+    pub modified_time: Option<SystemTime>,
+    pub hash: u64,
+    pub dependencies: Vec<ModulePath>,
+}
+
+pub type ModuleCacheMap = HashMap<ModulePath, ModuleCacheEntry>;
+
+#[derive(Clone, Debug)]
+pub struct ProgramsCacheEntry {
+    pub path: ModulePath,
+    pub compile_result: CompileResult<Programs>,
+}
+
+pub type ProgramsCacheMap = HashMap<ModulePath, ProgramsCacheEntry>;
+
 #[derive(Debug, Default)]
-pub struct QueryEngine {}
```

Comment:
This is a really nice start, but I think the query engine needs to be a lot more generic over queries.
Every query should be hashed and possibly have dependencies.